react/features/chat/middleware.js

```diff
@@ -27,49 +33,60 @@ declare var interfaceConfig : Object;
 MiddlewareRegistry.register(store => next => action => {
     switch (action.type) {
     case APP_WILL_MOUNT:
-        // Register the chat message sound on Web only because there's no chat
-        // on mobile.
-        typeof APP === 'undefined'
-            || store.dispatch(
+        store.dispatch(
                 registerSound(INCOMING_MSG_SOUND_ID, INCOMING_MSG_SOUND_FILE));
         break;
 
     case APP_WILL_UNMOUNT:
-        // Unregister the chat message sound on Web because it's registered
-        // there only.
-        typeof APP === 'undefined'
-            || store.dispatch(unregisterSound(INCOMING_MSG_SOUND_ID));
+        store.dispatch(unregisterSound(INCOMING_MSG_SOUND_ID));
         break;
 
     case CONFERENCE_JOINED:
-        typeof APP === 'undefined'
-            || _addChatMsgListener(action.conference, store);
+        _addChatMsgListener(action.conference, store);
         break;
 
-    case CONFERENCE_WILL_LEAVE:
-        store.dispatch(clearMessages());
-        break;
-
-    case SEND_MESSAGE:
-        if (typeof APP !== 'undefined') {
-            const { conference } = store.getState()['features/base/conference'];
+    case SEND_MESSAGE: {
+        const { conference } = store.getState()['features/base/conference'];
 
-            if (conference) {
-                const escapedMessage = UIUtil.escapeHtml(action.message);
+        if (conference) {
+            const escapedMessage = UIUtil.escapeHtml(action.message);
 
+            if (typeof APP !== 'undefined') {
                 APP.API.notifySendingChatMessage(escapedMessage);
-                conference.sendTextMessage(escapedMessage);
             }
+            conference.sendTextMessage(escapedMessage);
         }
         break;
     }
+    }
 
     return next(action);
 });
 
 /**
- * Registers listener for {@link JitsiConferenceEvents.MESSAGE_RECEIVED} which
- * will play a sound on the event, given that the chat is not currently visible.
+ * Set up state change listener to perform maintenance tasks when the conference
+ * is left or failed, e.g. clear messages or close the chat modal if it's left
+ * open.
+ */
+StateListenerRegistry.register(
```

Comment:
Unfortunately this is not enough: when a conference is joined while in another conference, the component is not unmounted so we need to cleanup in that case too.

Reply:
Added an equality check. We can generalise it later to handle all dialogs/modals